### ISSUE TITLE
[checkoutlinesufo] writes hashes even when not in -e mode

### ISSUE THREAD

**Issue by frankrolf** [adobe-type-tools/afdko#1090]
I noticed that `checkoutlinesufo` writes glyph hashes in the `/data` folder, even when not run in `-e` mode.
In the help text, I read
```
By default, a file containing compact descriptions
(a.k.a. hashes) of the glyphs is saved inside the UFO
on the first time the font is checked.
```
I would argue that this is not the idea of omitting the `-e` (for edit) argument – shouldn’t a UFO stay untouched if not explicitly requested?

You may convince me otherwise, but I do not understand the benefit of hashing the foreground glyphs without creating a processed layer.

**Comment by frankrolf** [comment]
One scenario where this may result in an unexpected outcome:
1. run `checkoutlinesufo font.ufo`:
glyphs are hashed (no processed layer is written)
2. run `checkoutlinesufo -e font.ufo`:
all glyphs are skipped (still no processed layer is written)

**Comment by frankrolf** [comment]
Another (related) issue:
1. run `checkoutlinesufo -e font.ufo`:
folder `glyphs.com.adobe.type.processedglyphs` is created
2. run `checkoutlinesufo -all font.ufo` (without `-e` argument): 
folder `glyphs.com.adobe.type.processedglyphs`is emptied except the `contents.plist` file, creating an invalid UFO.
<img width="532" alt="image" src="https://user-images.githubusercontent.com/2049645/74652131-a7722f80-5185-11ea-9bf9-d1a296de518b.png">

**Comment by josh-hadley** [comment]
Both of these behaviors seem like bugs to me. I'd be interested to get an opinion from @miguelsousa to confirm, but I think "no `-e`" should mean "no modifications", and `-all` should not generate an invalid UFO 😉 (I think the right approach would be to completely remove the entire `.processedglyphs` folder).

**Comment by miguelsousa** [comment]
I swear I've answered this kind of stuff before. I think that the `-h` option documents the tool well enough. In any case here's my take on this:

* **checkoutlinesufo** is a tool meant to analyze the glyphs' outlines and to provide feedback on them.
* The tool will also remove path overlaps when the `-e` option is used. The new paths are saved in the foreground layer if the `-w` option is also used, or in a new layer otherwise.
* Regardless of any options, the tool uses a cashing mechanism by default.

Therefore, "modifications" referes to changes related with glyph outlines rather than to changes to the UFO as a whole.

Since the hashing/cashing mechanism is on by default there will always be changes to the UFO whenever the tool is used on the font. It sounds like this issue is about proposing the hashes to be stored only when the `-e` option is used. I don't have a strong opinion against that but I'll say that, like Frank, I can probably come up with a sequence of steps and options that may produce unexpected results. Nonetheless, no matter what steps or options, the UFO must not end up corrupted.

**Comment by josh-hadley** [comment]
Thanks, Miguel. As someone not as familiar with the tool and its history, my impression from the the name _check_ outlinesufo (and its description summary in the help) is that its intended default behavior is to _look_ but not touch. I agree that the help generally does explain how the tool works, but the part about storing in the cache by default is somewhat buried in the description of the `--clear-hash-map` flag.

So maybe for that particular issue, all that needs to be done is explain that the cache will always be used, in the [tool-level description/help string](https://github.com/adobe-type-tools/afdko/blob/develop/python/afdko/checkoutlinesufo.py#L4)?

**Comment by frankrolf** [comment]
@miguelsousa We have indeed talked about this, in a slightly different, but related issue: https://github.com/adobe-type-tools/afdko/issues/283

I am not trying to bother anyone with pointing out these scenarios. Instead, the intent is to find reasons why the tool does not behave as I think it should. This issue does not imply “the tool is broken”, I want to point out possibly unexpected behavior, and give us a platform to discuss opinions.
(FWIW, I read the `-h` output and made several tests before bringing up these scenarios for discussion. The hashing decisions could indeed be described better – I volunteer to improve the help text as soon as we find a consensus).

My proposal:
- do not hash glyphs when not in `-e` mode.
The hash is created to speed up the creation of the corrected glyph (by skipping untouched ones). It does not make sense to store a hash if no corrected glyph is created.
- completely clear and re-write processed glyphs layer when using `-e --all`
- do not check existing hashes when not in `-e` mode
(this would leave hashes alone when just running `checkoutlinesufo` for the output)
- do not check existing hashes when in `-w` mode
(analog to #283, possibly already implemented.)

Again, please feel free to dismiss this proposal partly or completely.

**Comment by readroberts** [comment]
I agree that that checkoutlinesufo does not need to create hashes when not run in -e mode. Howe ver, it should respect and manage hashes that already exist. Specifically, if checkoutlinesufo has already been run, it should operate on the processed layer glyph rather than default layer glyph, but only if the hash is still valid for the default layer glyph. In the latter case, the default layer glyph has been edited since the hash was made, and checkoutlinesufo should still delete the process layer glyph and update the glyph hash.
